Source/WebCore/html/HTMLOptGroupElement.cpp

```diff
@@ -56,6 +56,37 @@ Ref<HTMLOptGroupElement> HTMLOptGroupElement::create(const QualifiedName& tagNam
     return adoptRef(*new HTMLOptGroupElement(tagName, document));
 }
 
+auto HTMLOptGroupElement::insertedIntoAncestor(InsertionType insertionType, ContainerNode& parentOfInsertedTree) -> InsertedIntoAncestorResult
+{
+    auto result = HTMLElement::insertedIntoAncestor(insertionType, parentOfInsertedTree);
+
+    if (!document().settings().htmlEnhancedSelectParsingEnabled())
+        return result;
+
+    if (auto maybeSelect = HTMLSelectElement::findAncestorSelectElement(*this, HTMLSelectElement::ExcludeOptGroup::Yes, &parentOfInsertedTree)) {
+        if (RefPtr select = *maybeSelect)
+            select->setRecalcListItems();
+    }
+
+    return result;
+}
+
+void HTMLOptGroupElement::removedFromAncestor(RemovalType removalType, ContainerNode& oldParentOfRemovedTree)
+{
+    HTMLElement::removedFromAncestor(removalType, oldParentOfRemovedTree);
+
+    if (!document().settings().htmlEnhancedSelectParsingEnabled())
+        return;
+
+    if (HTMLSelectElement::findAncestorSelectElement(*this, HTMLSelectElement::ExcludeOptGroup::Yes))
+        return;
+
+    if (auto maybeSelect = HTMLSelectElement::findAncestorSelectElement(oldParentOfRemovedTree, HTMLSelectElement::ExcludeOptGroup::Yes)) {
+        if (RefPtr select = *maybeSelect)
+            select->setRecalcListItems();
```

Comment:
Ditto. Do we need to do this work when the tree from which this subtree was removed was disconnected to begin with?